libraries/AP_IBus_Telem/AP_IBus_Telem.cpp

```diff
@@ -0,0 +1,674 @@
+            _incoming_command = c & 0x0f0;
+            _incoming_sensor_id = c & 0x0f;
+            _expected_checksum -= c;
+            _read_state = ReadState::READ_CHECKSUM_HIGH_BYTE;
+            break;
+
+        case ReadState::READ_CHECKSUM_HIGH_BYTE:
+            _incoming_checksum_high_byte = c;
+            _read_state = ReadState::READ_CHECKSUM_LOW_BYTE;
+            break;
+
+        case ReadState::READ_CHECKSUM_LOW_BYTE:
+            if (_expected_checksum == (c << 8) + _incoming_checksum_high_byte) {
+                handle_incoming_message();
+            }
+            _read_state = ReadState::DISCARD;
+            break;
+
+        case ReadState::DISCARD:
+            break;
+        }
+    }
+}
+
+void AP_IBus_Telem::handle_incoming_message()
+{
+    // Sensor 0 is reserved for the transmitter, so if for some reason it's requested we should ignore it
+    if (_incoming_sensor_id == 0) {
+        return;
+    }
+
+    // If we've reached the end of our sensor list, we shouldn't respond; this tells the receiver that there
+    // are no more sensors to discover.
+    if (_incoming_sensor_id > ARRAY_SIZE(_sensors)) {
+        return;
+    }
+
+    const auto *sensor_definition = &_sensors[_incoming_sensor_id - 1];
+
+    switch (_incoming_command) {
+    case PROTOCOL_COMMAND_DISCOVER:
+        handle_discover_command();
+        break;
+
+    case PROTOCOL_COMMAND_TYPE:
+        handle_type_command(*sensor_definition);
+        break;
+
+    case PROTOCOL_COMMAND_VALUE:
+        handle_value_command(*sensor_definition);
+        break;
+    }
+}
+
+/* A discovery query has the following format:
+   * 0x04: Message length
+   * 0x81: 0x80 for discovery + 0x01 for sensor ID 1
+   * 0x7A: Checksum low byte
+   * 0xFF: Checksum high byte
+   Checksums are 0xFFFF minus the sum of the previous bytes
+
+   To acknowledge a discovery query, we echo the command back.
+ */
+void AP_IBus_Telem::handle_discover_command()
+{
+    struct protocol_command_discover_response_t {
+        uint8_t command_length;
+        uint8_t address;
+        uint16_t checksum;
+    } packet {
+        PROTOCOL_FOUR_LENGTH,
+        (uint8_t)(PROTOCOL_COMMAND_DISCOVER | _incoming_sensor_id)
+    };
+    populate_checksum((uint8_t*)&packet, sizeof(packet));
+    _port->write((uint8_t*)&packet, sizeof(packet));
+}
+
+/* A type query has the following format:
+   * 0x04: Message length
+   * 0x91: 0x90 for type + 0x01 for sensor ID 1
+   * 0x6A: Checksum low byte
+   * 0xFF: Checksum high byte
+   Checksums are 0xFFFF minus the sum of the previous bytes
+
+   To respond to a type query, we send:
+   * 0x06: Message length
+   * 0x91: 0x90 for type + 0x01 for sensor ID 1
+   * 0x03: Sensor type, eg 0x03 for external voltage
+   * 0x02: Sensor length (2 or 4 bytes)
+   * 0x63: Checksum low byte
+   * 0xFF: Checksum high byte
+   Checksums are 0xFFFF minus the sum of the previous bytes
+ */
+void AP_IBus_Telem::handle_type_command(SensorDefinition sensor)
```

Comment:
```suggestion
void AP_IBus_Telem::handle_type_command(const SensorDefinition &sensor)
```

& avoids putting data onto the stack, and `const` is just generally useful for keeping logic correct.

similarly elsewhere